ocs_ci/ocs/resources/cloud_manager.py

```diff
@@ -0,0 +1,225 @@
+        pass
+
+    def verify_uls_exists(self, uls_name):
+        pass
+
+    @abstractmethod
+    def internal_create_uls(self, name):
+        pass
+
+    @abstractmethod
+    def internal_delete_uls(self, name):
+        pass
+
+
+class S3Client(CloudClient):
+    """
+        Implementation of a S3 Client using the S3 API
+    """
+
+    def __init__(self, key_id, access_key, endpoint="https://s3.amazonaws.com", verify=True,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.client = boto3.resource(
+            's3', verify=verify, endpoint_url=endpoint,
+            aws_access_key_id=key_id,
+            aws_secret_access_key=access_key
+        )
+        bs_secret_data = templating.load_yaml(constants.MCG_BACKINGSTORE_SECRET_YAML)
+        bs_secret_data['metadata']['name'] += f'-client-secret'
+        bs_secret_data['metadata']['namespace'] = config.ENV_DATA['cluster_namespace']
+        bs_secret_data['data']['AWS_ACCESS_KEY_ID'] = base64.urlsafe_b64encode(
+            self.aws_access_key_id.encode('UTF-8')
+        ).decode('ascii')
+        bs_secret_data['data']['AWS_SECRET_ACCESS_KEY'] = base64.urlsafe_b64encode(
+            self.aws_access_key.encode('UTF-8')
+        ).decode('ascii')
+        self.secret = create_resource(**bs_secret_data)
+
+    def internal_create_uls(self, name, region=None):
+        """
+            Creates the Underlying Storage using the S3 API
+        """
+        if region is None:
+            self.client.create_bucket(Bucket=name)
+        else:
+            self.client.create_bucket(
+                Bucket=name,
+                CreateBucketConfiguration={
+                    'LocationConstraint': region
+                }
+            )
+
+    def internal_delete_uls(self, name):
+        """
+            Deletes the Underlying Storage using the S3 API
+        """
+        self.client.meta.client.delete_bucket_policy(
+            Bucket=name
+        )
+        for _ in range(10):
+            try:
+                self.client.Bucket(name).objects.all().delete()
+                self.client.Bucket(name).delete()
+                break
+            except ClientError:
+                logger.info(f'Deletion of ULS {name} failed. Retrying...')
+                sleep(3)
+
+    def get_all_uls_names(self):
+        """
+            Returns:
+                set: A set of all bucket names
+
+        """
+        return {bucket.name for bucket in self.client.buckets.all()}
+
+    def verify_uls_exists(self, uls_name):
+        """
+           Verifies whether a uls with the given uls_name exists
+           Args:
+               uls_name : The uls name to be verified
+
+           Returns:
+                 bool: True if uls exists, False otherwise
+
+        """
+        try:
+            self.client.head_bucket(Bucket=uls_name)
+            logger.info(f"{uls_name} exists")
+            return True
+        except ClientError:
```

Comment:
isn't ClientError too generic?

Reply:
from the boto3 docs:
``` except botocore.exceptions.ClientError as e:
    # If a client error is thrown, then check that it was a 404 error.
    # If it was a 404 error, then the bucket does not exist.
    error_code = e.response['Error']['Code']
    if error_code == '404':
        exists = False
```
i can add the error code as shown here but didn't see this implementation in other places in the code